### DISCUSSION TITLE
Testing a normal provider in the same way as I test a notifier provider

### DISCUSSION THREAD

**Discussion by egonm12** [rrousselGit/riverpod#2454] [Q&A] [answered]
I am using a `ProviderContainer` to test the behaviour of a simple `Provider`. I would like to verify all calls made to my listener, in a similar way as I test a notifier provider. I read in the docs that listeners will automatically be removed when the provider rebuilds (such as when a provider listened with `watch` changes), so maybe that is causing the behavior for the normal provider to be different since it depends on another provider with `watch`.

Consider these providers:
```
@riverpod
bool isCounterOdd(IsCounterOddRef ref) {
  final counter = ref.watch(counterProvider);

  return counter.isOdd;
}

@riverpod
class Counter extends _$Counter {
  @override
  int build() => 0;

  void increment() {
    state = state + 1;
  }
}
```

This test succeeds:
```
test('some test description', () async {
  final listener = Listener<bool>();
  final container = ProviderContainer()
    ..listen(
      isCounterOddProvider,
      listener.call,
      fireImmediately: true,
    );

  container.read(counterProvider.notifier).increment();
  container.read(isCounterOddProvider);
  container.read(counterProvider.notifier).increment();
  container.read(isCounterOddProvider);

  verifyInOrder([
    () => listener.call(null, false),
    () => listener.call(false, true),
    () => listener.call(true, false),
  ]);
  verifyNoMoreInteractions(listener);
});
```

When I remove `container.read(isCounterOddProvider);` it only registers one call due to `fireImmediately: true`. I would expect that the `watch` in my `isCounterOddProvider` would ensure the listener to get triggered, but apparently that is not the case. Is there another way to achieve this without manually reading the provider after every interaction? It feels really cumbersome at the moment to achieve this. I'd like to test it similar to a notifier provider.

**Comment by rrousselGit** [comment, marked as answer]
Providers rebuild asynchronously.

You need to call "container.pump()" to allow providers to naturally rebuild

**Reply by egonm12** [reply]
@rrousselGit if I should avoid it, what do you propose as an alternative? I see in the riverpod codebase tests that do use 'container.pump()', so I am a bit confused.

**Reply by rrousselGit** [reply]
Oh sorry, I meant `call` not `avoid`. Do use `container.pump()`

**Reply by egonm12** [reply]
Thanks, that seems to do the trick 👍🏻
